File: backend/app/api/mtd.py
import uuid
import json
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session, selectinload
from app.core.database import get_db
from app.models.mtd_record import MTDRecord
from app.models.order import Order
from app.lib.producer_assignment import (
    canonical_producer_assignment_key,
    normalize_producer_key,
    resolve_producer_by_assignment_key,
)
from app.schemas.mtd_record import (
    MTDRecordSchema,
    MTDRecordCreateSchema,
    MTDRecordUpdateSchema,
    ManualScheduleCreateSchema,
)

router = APIRouter()

def is_valid_uuid(val: str) -> bool:
    try:
        uuid.UUID(val)
        return True
    except (ValueError, TypeError, AttributeError):
        return False

def _find_mtd(db: Session, mtd_id: str) -> MTDRecord | None:
    if is_valid_uuid(mtd_id):
        parsed_uuid = uuid.UUID(mtd_id)
        mtd = (
            db.query(MTDRecord)
            .filter(
                (MTDRecord.id == parsed_uuid)
                | (MTDRecord.order_id == parsed_uuid)
                | (MTDRecord.legacy_id == mtd_id)
            )
            .first()
        )
        if mtd:
            return mtd
    return db.query(MTDRecord).filter(MTDRecord.legacy_id == mtd_id).first()

@router.get("/mtd", response_model=List[MTDRecordSchema])
def get_mtd_records(
    form_type: str | None = None,
    cheer_form_subtype: str | None = None,
    category: str | None = None,
    status: str | None = None,
    db: Session = Depends(get_db)
):
    # Eager-load the producer relationship so serializing assigned_producer for
    # every row does NOT fire one query per record (N+1). This collapses ~248
    # extra round-trips to Supabase into a single batched query.
    query = db.query(MTDRecord).options(selectinload(MTDRecord.assigned_producer))
    if cheer_form_subtype and cheer_form_subtype != "all":
        query = query.join(Order, MTDRecord.order_id == Order.id).filter(Order.cheer_form_subtype == cheer_form_subtype)
    elif form_type:
        query = query.join(Order, MTDRecord.order_id == Order.id).filter(Order.form_type == form_type)
    if category and category != "All":
        query = query.filter(MTDRecord.category == category)
    if status:
        query = query.filter(MTDRecord.status == status)
    return query.all()

@router.get("/mtd/{mtd_id}", response_model=MTDRecordSchema)
def get_mtd_record(mtd_id: str, db: Session = Depends(get_db)):
    mtd = _find_mtd(db, mtd_id)
    if not mtd:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="MTD Record not found")
    return mtd

from app.api.auth import require_full_access

@router.post("/mtd", response_model=MTDRecordSchema, status_code=status.HTTP_201_CREATED)
def create_mtd_record(payload: MTDRecordCreateSchema, db: Session = Depends(get_db), _: None = Depends(require_full_access)):
    data = payload.model_dump()
    order_id = data.get("order_id")

    # If an MTD record already exists for this order_id, update it instead of duplicating
    existing_mtd = None
    if order_id:
        existing_mtd = _find_mtd(db, str(order_id))

    if existing_mtd:
        assigned_prod_str = data.pop("assigned_producer", None)
        if assigned_prod_str:
            producer = resolve_producer_by_assignment_key(db, assigned_prod_str)
            existing_mtd.assigned_producer_id = producer.id if producer else None
            existing_mtd.editor_initials = (
                canonical_producer_assignment_key(producer)
                if producer
                else assigned_prod_str.strip().upper()
            )
        for field, value in data.items():
            if value is not None:
                setattr(existing_mtd, field, value)
        db.commit()
        db.refresh(existing_mtd)
        return existing_mtd

    assigned_prod_str = data.pop("assigned_producer", None)
    assigned_producer_id = None
    editor_initials = None
    if assigned_prod_str:
        producer = resolve_producer_by_assignment_key(db, assigned_prod_str)
        if producer:
            assigned_producer_id = producer.id
            editor_initials = canonical_producer_assignment_key(producer)
        else:
            editor_initials = assigned_prod_str.strip().upper()

    mtd = MTDRecord(
        **data,
        assigned_producer_id=assigned_producer_id,
        editor_initials=editor_initials,
    )
    db.add(mtd)
    db.commit()
    db.refresh(mtd)
    return mtd

@router.patch("/mtd/{mtd_id}", response_model=MTDRecordSchema)
def update_mtd_record(mtd_id: str, payload: MTDRecordUpdateSchema, db: Session = Depends(get_db), _: None = Depends(require_full_access)):
    mtd = _find_mtd(db, mtd_id)
    if not mtd:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="MTD Record not found")

    update_data = payload.model_dump(exclude_unset=True)

    if "assigned_producer" in update_data:
        assigned_prod_str = update_data.pop("assigned_producer")
        previous_key = None
        if mtd.assigned_producer:
            previous_key = canonical_producer_assignment_key(mtd.assigned_producer)
        elif mtd.editor_initials:
            previous_key = mtd.editor_initials.strip().upper()

        if assigned_prod_str:
            producer = resolve_producer_by_assignment_key(db, assigned_prod_str)
            mtd.assigned_producer_id = producer.id if producer else None
            mtd.editor_initials = (
                canonical_producer_assignment_key(producer)
                if producer
                else assigned_prod_str.strip().upper()
            )
        else:
            mtd.assigned_producer_id = None
            if (
                previous_key
                and mtd.editor_initials
                and normalize_producer_key(mtd.editor_initials) == normalize_producer_key(previous_key)
            ):
                mtd.editor_initials = None

    for key, value in update_data.items():
        # Text columns that store JSON dicts must be serialised to string
        if isinstance(value, (dict, list)):
            value = json.dumps(value)
        setattr(mtd, key, value)

    # Sync fields to linked Order if present
    if mtd.order_id:
        linked_order = db.query(Order).filter(Order.id == mtd.order_id).first()
        if linked_order:
            if "contact_name" in update_data:
                linked_order.contact_name = mtd.contact_name
                linked_order.customer_name = mtd.contact_name
            if "program_name" in update_data:
                linked_order.program_name = mtd.program_name
            if "package" in update_data:
                linked_order.package = mtd.package
            if "music_theme" in update_data:
                linked_order.music_theme = mtd.music_theme
            if "price" in update_data:
                linked_order.price = mtd.price
            if "price_compliance" in update_data:
                linked_order.price_compliance = mtd.price_compliance
            if "is_reassigned" in update_data:
                linked_order.is_reassigned = mtd.is_reassigned
            if "order_status" in update_data:
                linked_order.order_status = mtd.order_status
            if "missing_data_email_sent_at" in update_data:
                linked_order.missing_data_email_sent_at = mtd.missing_data_email_sent_at
            if "in_mtd" in update_data:
                # Keep Order.status aligned with MTD board membership so Move to
                # Orders / Move to MTD survive reloads.
                if mtd.in_mtd:
                    linked_order.status = "in_mtd"
                elif linked_order.status == "in_mtd":
                    linked_order.status = "active"
            if "status" in update_data and mtd.status == "completed":
                linked_order.status = "completed"

    db.commit()
    db.refresh(mtd)
    return mtd


@router.post("/mtd/manual-schedule", response_model=MTDRecordSchema, status_code=status.HTTP_201_CREATED)
def create_manual_schedule_entry(
    payload: ManualScheduleCreateSchema,
    db: Session = Depends(get_db),
    _: None = Depends(require_full_access)
):
    data = payload.model_dump()
    assigned_prod_str = data.pop("assigned_producer", None)
    producer = resolve_producer_by_assignment_key(db, assigned_prod_str) if assigned_prod_str else None

    category = data.get("category") or "Cheer"
    section = "DANCE MUSIC" if category == "Dance" else "CHEERLEADING MUSIC"

    program_name = (data.get("program_name") or "").strip() or "Manual Schedule Entry"
    contact_name = (data.get("contact_name") or "").strip() or "N/A"
    package = (data.get("package") or "").strip() or "Standard"

    mtd = MTDRecord(
        order_id=None,
        section=section,
        category=category,
        program_name=program_name,
        contact_name=contact_name,
        package=package,
        price=0.0,
        price_compliance="compliant",
        invoice="",
        mix_start_date=data.get("mix_start_date"),
        mix_end_date=data.get("mix_end_date"),
        assigned_producer_id=producer.id if producer else None,
        editor_initials=canonical_producer_assignment_key(producer) if producer else (assigned_prod_str or "UNASSIGNED"),
        editor_request="FA",
        status="active",
        in_mtd=False,
        is_manual_schedule_entry=True,
        routine_notes=data.get("routine_notes"),
        music_affiliate=data.get("music_affiliate"),
        time_length_of_mix=data.get("time_length_of_mix"),
        song_list_suggestions=data.get("song_list_suggestions"),
        custom_voiceovers=data.get("custom_voiceovers"),
        eight_count_sheet=data.get("eight_count_sheet") or "NEED CS",
        have_songs="NEED SONGS",
        needs_attention=False,
    )

    db.add(mtd)
    db.commit()
    db.refresh(mtd)
    return mtd

File: .test-build/lib/api/mtd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformMTDRecord = transformMTDRecord;
exports.fetchMTDRecordsApi = fetchMTDRecordsApi;
exports.createMTDRecordApi = createMTDRecordApi;
exports.updateMTDRecordApi = updateMTDRecordApi;
const client_1 = require("./client");
function transformMTDRecord(bm) {
    return {
        id: bm.legacy_id || bm.id,
        legacyId: bm.legacy_id || undefined,
        uuid: bm.id,
        orderId: bm.order_id || undefined,
        section: bm.section || "CHEERLEADING MUSIC",
        assignedProducer: bm.assigned_producer && bm.assigned_producer !== "FA" && bm.assigned_producer !== "NA" ? bm.assigned_producer : null,
        category: bm.category || "Cheer",
        editorRequest: bm.editor_request || "FA",
        contactName: bm.contact_name || "",
        editorInitials: bm.editor_initials || bm.assigned_producer || "",
        programName: bm.program_name || "",
        package: bm.package || "",
        musicTheme: bm.music_theme || "",
        price: bm.price ?? 0,
        priceCompliance: bm.price_compliance || "compliant",
        invoice: bm.invoice || "",
        mixStartDate: bm.mix_start_date || "",
        mixEndDate: bm.mix_end_date || undefined,
        waitingOn: bm.waiting_on || null,
        eightCountSheet: bm.eight_count_sheet || "",
        haveSongs: bm.have_songs || "",
        needsAttention: Boolean(bm.needs_attention),
        status: bm.status || "active",
        recordStatus: bm.record_status || undefined,
        inMTD: Boolean(bm.in_mtd ?? bm.inMTD),
        inPayroll: Boolean(bm.in_payroll),
        completedAt: bm.completed_at || undefined,
        hasRallyMix: Boolean(bm.has_rally_mix ?? bm.hasRallyMix),
        hasExtend8ctAddon: Boolean(bm.has_extend_8ct_addon ?? bm.hasExtend8ctAddon),
        hasProcessing8ctSheetsAddon: Boolean(bm.has_processing_8ct_sheets_addon ?? bm.hasProcessing8ctSheetsAddon),
        systemCalculatedCustomerPrice: bm.system_calculated_customer_price ?? null,
        finalCustomerPrice: bm.final_customer_price ?? null,
        finalCustomerPriceOverridden: Boolean(bm.final_customer_price_overridden),
        pricingBreakdown: bm.pricing_breakdown || null,
        rateUsed: bm.rate_used ?? null,
        rateSource: bm.rate_source || null,
        producerPayout: bm.producer_payout ?? null,
        sltPortion: bm.slt_portion ?? null,
        payrollFinalized: Boolean(bm.payroll_finalized),
        payrollBreakdown: bm.payroll_breakdown || null,
    };
}
async function fetchMTDRecordsApi() {
    const backendMtd = await client_1.apiClient.get("/api/mtd");
    return backendMtd.map(transformMTDRecord);
}
async function createMTDRecordApi(record) {
    const payload = {
        order_id: record.orderId || null,
        section: record.section || "CHEERLEADING MUSIC",
        category: record.category || "Cheer",
        contact_name: record.contactName || "",
        program_name: record.programName || "",
        package: record.package || "",
        price: record.price ?? 0,
        music_theme: record.musicTheme || null,
        editor_request: record.editorRequest || "FA",
        assigned_producer: record.assignedProducer || null,
        invoice: record.invoice || "",
        mix_start_date: record.mixStartDate || null,
        mix_end_date: record.mixEndDate || null,
        record_status: record.recordStatus || (record.assignedProducer ? "Ongoing" : null),
        eight_count_sheet: record.eightCountSheet || "NEED CS",
        have_songs: record.haveSongs || "NEED SONGS",
        needs_attention: record.needsAttention ?? false,
        status: record.status || "active",
        in_mtd: Boolean(record.inMTD),
        has_rally_mix: Boolean(record.hasRallyMix),
        has_extend_8ct_addon: Boolean(record.hasExtend8ctAddon),
        has_processing_8ct_sheets_addon: Boolean(record.hasProcessing8ctSheetsAddon),
    };
    const res = await client_1.apiClient.post("/api/mtd", payload);
    return transformMTDRecord(res);
}
async function updateMTDRecordApi(id, patch) {
    const payload = {};
    if (patch.section !== undefined)
        payload.section = patch.section;
    if (patch.assignedProducer !== undefined)
        payload.assigned_producer = patch.assignedProducer;
    if (patch.category !== undefined)
        payload.category = patch.category;
    if (patch.editorRequest !== undefined)
        payload.editor_request = patch.editorRequest;
    if (patch.contactName !== undefined)
        payload.contact_name = patch.contactName;
    if (patch.editorInitials !== undefined)
        payload.editor_initials = patch.editorInitials;
    if (patch.programName !== undefined)
        payload.program_name = patch.programName;
    if (patch.package !== undefined)
        payload.package = patch.package;
    if (patch.musicTheme !== undefined)
        payload.music_theme = patch.musicTheme;
    if (patch.price !== undefined)
        payload.price = patch.price;
    if (patch.priceCompliance !== undefined)
        payload.price_compliance = patch.priceCompliance;
    if (patch.invoice !== undefined)
        payload.invoice = patch.invoice;
    if (patch.mixStartDate !== undefined)
        payload.mix_start_date = patch.mixStartDate || null;
    if (patch.mixEndDate !== undefined)
        payload.mix_end_date = patch.mixEndDate || null;
    if (patch.waitingOn !== undefined)
        payload.waiting_on = patch.waitingOn;
    if (patch.eightCountSheet !== undefined)
        payload.eight_count_sheet = patch.eightCountSheet;
    if (patch.haveSongs !== undefined)
        payload.have_songs = patch.haveSongs;
    if (patch.needsAttention !== undefined)
        payload.needs_attention = patch.needsAttention;
    if (patch.status !== undefined)
        payload.status = patch.status;
    if (patch.recordStatus !== undefined)
        payload.record_status = patch.recordStatus;
    if (patch.inMTD !== undefined)
        payload.in_mtd = patch.inMTD;
    if (patch.inPayroll !== undefined)
        payload.in_payroll = patch.inPayroll;
    if (patch.completedAt !== undefined)
        payload.completed_at = patch.completedAt;
    if (patch.hasRallyMix !== undefined)
        payload.has_rally_mix = patch.hasRallyMix;
    if (patch.hasExtend8ctAddon !== undefined)
        payload.has_extend_8ct_addon = patch.hasExtend8ctAddon;
    if (patch.hasProcessing8ctSheetsAddon !== undefined)
        payload.has_processing_8ct_sheets_addon = patch.hasProcessing8ctSheetsAddon;
    if (patch.systemCalculatedCustomerPrice !== undefined) {
        payload.system_calculated_customer_price = patch.systemCalculatedCustomerPrice;
    }
    if (patch.finalCustomerPrice !== undefined) {
        payload.final_customer_price = patch.finalCustomerPrice;
    }
    if (patch.finalCustomerPriceOverridden !== undefined) {
        payload.final_customer_price_overridden = patch.finalCustomerPriceOverridden;
    }
    if (patch.rateUsed !== undefined)
        payload.rate_used = patch.rateUsed;
    if (patch.rateSource !== undefined)
        payload.rate_source = patch.rateSource;
    if (patch.producerPayout !== undefined)
        payload.producer_payout = patch.producerPayout;
    if (patch.sltPortion !== undefined)
        payload.slt_portion = patch.sltPortion;
    if (patch.payrollFinalized !== undefined) {
        payload.payroll_finalized = patch.payrollFinalized;
    }
    if (patch.payrollBreakdown !== undefined) {
        payload.payroll_breakdown = patch.payrollBreakdown;
    }
    try {
        const res = await client_1.apiClient.patch(`/api/mtd/${id}`, payload);
        return transformMTDRecord(res);
    }
    catch (err) {
        if (err instanceof client_1.ApiClientError) {
            console.warn(`MTD Record ${id} update not persisted to backend (${err.message}). Local update retained.`);
            return { id, ...patch };
        }
        throw err;
    }
}
